### DISCUSSION TITLE
How to use Babel Plugin to manipulate AST?

### DISCUSSION THREAD

**Discussion by liuliangsir** [babel/babel#11613] [Q&A] [answered]
As the above title says, I would like to use  Babel Plugin to add a custom property(like start or end) on AST generated by using [transformAsync](https://babeljs.io/docs/en/babel-core#transformasync). Is there any information about how to resolve the issue above?

**Comment by xtuc** [comment]
Could you please tell us more about your use case?

If you want to manipulate every node in an AST you could use something like:
```js
traverse(ast, {
  enter(path) {
    path.node.something = 1;
  }
  exit(path) {
    path.node.something = 2;
  }
});

```

**Reply by khushal87** [reply]
Hey I am getting 
```
traverse(ast, {
^

TypeError: traverse is not a function
```
Can you help me how can I solve it? @xtuc

**Comment by liuliangsir** [comment]
@xtuc Okay, I want to add a custom property(such as uuid) on AST object, which is at the same level as `start` property in the following.

![image](https://user-images.githubusercontent.com/8049878/82896092-ec2d1400-9f87-11ea-9fdb-9142a62962c9.png)

**Comment by nicolo-ribaudo** [comment, marked as answer]
@liuliangsir To expand @xtuc's answer, here is an example of a plugin adding an `uid` property to every node: https://astexplorer.net/#/gist/09b316493a94ae3ccb0e23e209344390/8b71a923763676342ed9ec1e829daaa40edeac5b

**Comment by liuliangsir** [comment]
@nicolo-ribaudo thx, It saves my life.
